rust/connlib/connection/src/pool.rs

```diff
@@ -0,0 +1,632 @@
+use std::time::{Duration, Instant};
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    net::SocketAddr,
+    sync::Arc,
+};
+use str0m::ice::{IceAgent, IceAgentEvent, IceCreds};
+use str0m::net::{Protocol, Receive};
+use str0m::{Candidate, StunMessage};
+
+use crate::index::IndexLfsr;
+use crate::IpPacket;
+
+// Note: Taken from boringtun
+const HANDSHAKE_RATE_LIMIT: u64 = 100;
+
+const MAX_UDP_SIZE: usize = (1 << 16) - 1;
+
+/// Manages a set of wireguard connections for a server.
+pub type ServerConnectionPool<TId> = ConnectionPool<Server, TId>;
+/// Manages a set of wireguard connections for a client.
+pub type ClientConnectionPool<TId> = ConnectionPool<Client, TId>;
+
+pub enum Server {}
+pub enum Client {}
+
+pub struct ConnectionPool<T, TId> {
+    private_key: StaticSecret,
+    index: IndexLfsr,
+    rate_limiter: Arc<RateLimiter>,
+    local_interfaces: HashSet<SocketAddr>,
+    buffered_transmits: VecDeque<Transmit>,
+
+    next_rate_limiter_reset: Option<Instant>,
+
+    initial_connections: HashMap<TId, InitialConnection>,
+    negotiated_connections: HashMap<TId, Connection>,
+    pending_events: VecDeque<Event<TId>>,
+
+    buffer: Box<[u8; MAX_UDP_SIZE]>,
+
+    marker: PhantomData<T>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Unknown interface")]
+    UnknownInterface,
+    #[error("Failed to decapsulate: {0:?}")] // TODO: Upstream an std::error::Error impl
+    Decapsulate(boringtun::noise::errors::WireGuardError),
+    #[error("Failed to encapsulate: {0:?}")]
+    Encapsulate(boringtun::noise::errors::WireGuardError),
+    #[error("Unmatched packet")]
+    UnmatchedPacket,
+    #[error("Not connected")]
+    NotConnected,
+}
+
+impl<T, TId> ConnectionPool<T, TId>
+where
+    TId: Eq + Hash + Copy + fmt::Display,
+{
+    pub fn new(private_key: StaticSecret) -> Self {
+        let public_key = &(&private_key).into();
+        Self {
+            private_key,
+            marker: Default::default(),
+            index: IndexLfsr::default(),
+            rate_limiter: Arc::new(RateLimiter::new(public_key, HANDSHAKE_RATE_LIMIT)),
+            local_interfaces: HashSet::default(),
+            buffered_transmits: VecDeque::default(),
+            next_rate_limiter_reset: None,
+            negotiated_connections: HashMap::default(),
+            pending_events: VecDeque::default(),
+            initial_connections: HashMap::default(),
+            buffer: Box::new([0u8; MAX_UDP_SIZE]),
+        }
+    }
+
+    pub fn add_local_interface(&mut self, local_addr: SocketAddr) {
+        self.local_interfaces.insert(local_addr);
+
+        // TODO: Add host candidate to all existing connections here.
+    }
+
+    pub fn add_remote_candidate(&mut self, id: TId, candidate: String) {
+        let candidate = match Candidate::from_sdp_string(&candidate) {
+            Ok(c) => c,
+            Err(e) => {
+                tracing::debug!("Failed to parse candidate: {e}");
```

Comment:
I think `warn` is more adequate here

Reply:
or maybe return a `Result` for `add_remote_candidate`?

Reply:
Resolving this in favor of the discussion above.